### DISCUSSION TITLE
useFormState - Prevent page refreshes upon submission!?

### DISCUSSION THREAD

**Discussion by Micrified** [vercel/next.js#69069] [Help]
### Summary

# Context

I have a form that sits within a page requiring authentication. My `layout.js` checks and **redirects** the user before the page loads if they're not logged in. However, the user may also sit on the page a long time, and become logged out. When this happens, I have written my own redirect to the login with a few query parameters. 

```
'use client'

const defaultState = { ok: false, status: 0, err: "", data: {}};

export default function Page() {
  const router = useRouter();
  const [state, action] = useFormState(MyAct, defaultState);
  const pathname = usePathname();

  useEffect(() => {
      if (state.ok) {
        router.push(...);
      } else {
        router.push(`/login?target=${pathname}`);
      }
    }
  }, [state]);

  return ( <form action={action}> ... </form>);
}
```

# Problem 

The problem is that I never get to execute my redirect. Instead:
1. The user clicks `Submit` in the form
2. The form is submitted, and a refresh occurs
3. The page is redirected (by the layout.js) back to the login page

This isn't what I want. I'd like to be able to overrule this behaviour. Normally, a `e.preventDefault` is possible in a custom form handler. However, that behaviour is now handled somewhere within the server action. How can I prevent this refresh from occurring? 

---

### Similar questions

* https://github.com/vercel/next.js/discussions/66401: Asks why `e.preventDefault()` _doesn't_ need to be called, so somewhat off topic

### Additional information

```js
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.2.0: Wed Nov 15 21:54:51 PST 2023; root:xnu-10002.61.3~2/RELEASE_ARM64_T6030
  Available memory (MB): 18432
  Available CPU cores: 11
Binaries:
  Node: 18.20.2
  npm: 10.8.1
  Yarn: N/A
  pnpm: N/A
Relevant Packages:
  next: 14.2.5 // Latest available version is detected (14.2.5).
  eslint-config-next: 14.2.5
  react: 18.3.1
  react-dom: 18.3.1
  typescript: 5.4.5
Next.js Config:
  output: N/A
```

**Comment by Micrified** [comment]
⁇

Why not re-direct from _within_ the server-action? Then, you might preempt the form behaviour? Two reasons:
1. It isn't clear which occurs first. And I'd rather disable that form behaviour.
2. Redirecting in server-actions doesn't work anyways. See [this Github Discussion](https://github.com/vercel/next.js/issues/58263) indicating that it is broken in `NextJS 14.2.X`. 
---

**Edit**: Follow-up, turns out if you embed a server-action **within the same page** (following guidance from [the NextJS subreddit](https://www.reddit.com/r/nextjs/comments/18pirsd/redirect_error_in_nextjs_server_action/)), you kind of can of get the re-direct to work, but I haven't been able to take advantage of this because my server-action makes use of cookies. You may not import cookies from `next/headers` within a client component. Foiled again 😔

**Comment by matija2209** [comment]
No matter what I try my page always gets reloaded when a form via useFomrState is submitted.
